### PR TITLE
[MCHECKSTYLE-401] - NullPointerException in Violation class

### PR DESCRIPTION
**Pull request by stevenschenk**
Fixes [MCHECKSTYLE-401](https://issues.apache.org/jira/browse/MCHECKSTYLE-401). 

Both `file` and `line` field are nullable in the constructor. Equals method now uses `Objects.equals()` for those fields to handle the nullability. Also added a test for the `equals()` method.

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MCHECKSTYLE) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MCHECKSTYLE-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MCHECKSTYLE-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the integration tests successfully (`mvn -Prun-its clean verify`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by bmarwell** [comment]
@stevenschenk the checkstyle API states that file and line can never be null. So I expect you are fixing a problem which can never occur, only in tests.
I mean, how can a violation not be on a line in a file? 

That said, I have no objections merging it anyway.

**Comment by stevenschenk** [comment]
> @stevenschenk the checkstyle API states that file and line can never be null. So I expect you are fixing a problem which can never occur, only in tests.
> I mean, how can a violation not be on a line in a file?
> 
> That said, I have no objections merging it anyway.

My bad. Was not listed in the ticket, but I did not verify it either. If it cannot be null, should we do a `requiresNonNull()` check in the constructor?

File: src/main/java/org/apache/maven/plugins/checkstyle/Violation.java [hunk 1]
```diff
@@ -196,10 +196,10 @@ public boolean equals( Object other )
       return false;
     }
     Violation violation = ( Violation ) other;
-    return line.equals( violation.line )
+    return Objects.equals( line, violation.line )
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/checkstyle/Violation.java:199]: changes should have unit tests

Comment (stevenschenk) [reply src/main/java/org/apache/maven/plugins/checkstyle/Violation.java:199]: I added a test for the equals method. Could you elaborate on what you specifically want to be tested?

Comment (bmarwell) [reply src/main/java/org/apache/maven/plugins/checkstyle/Violation.java:199]: Can confirm it is running through this line.
2021-06-04T22:28:41_0001_screenshot

**Comment by stevenschenk** [comment]
> > @stevenschenk the checkstyle API states that file and line can never be null. So I expect you are fixing a problem which can never occur, only in tests.
> > I mean, how can a violation not be on a line in a file?
> > That said, I have no objections merging it anyway.
> 
> My bad. Was not listed in the ticket, but I did not verify it either. I checked the constructor and there was no null check on these fields, so I assumed they were nullable. If the API states they cannot be null, should we do a `requiresNonNull()` check in the constructor?

@bmarwell So I added a `requiresNonNull()` check for `file` and `line` fields in the constructor of `Violation`, however it causes other tests to fail (For example: `CheckstyleViolationCheckMojoTesttestDefaultConfig()`. One of the test cases in that test is a violation for a missing file. When creating the violation it provides the "missing file" as file input to the constructor. As the missing file could not be resolved, resulting in a null value, it seems violation could be created with null value for `file`. This seems contradicting with the API. Is this a test case that can never occur, or could violations actually exist without a file?

**Review by elharo — APPROVED** [review]

**Comment by elharo** [comment]
The checkstyle API is not part of our API and we should not rely on the promises it makes. Trust but verify. This is public API as @stevenschenk points out there is no null check, and it is possible for someone to pass null for these values. Therefore the equals method should handle that case.

**Comment by elharo** [comment]
jenkins: https://ci-builds.apache.org/job/Maven/job/maven-box/job/maven-checkstyle-plugin/job/MCHECKSTYLE-401/

**Comment by jira-importer** [comment]
Resolve #314

**Pull request merged**